### DISCUSSION TITLE
InputModel for inserting a record with required fields

### DISCUSSION THREAD

**Discussion by tqwewe** [SeaQL/sea-orm#96] [Ideas]
It doesn't look like SeaORM currently stores information about whether an SQL column has a default value.
Storing this could be useful for generating a new struct perhaps called `InputModel` which wraps each column in an `Option<T>` **except** those that are required (non-nullable) and do not have a default value. Auto-generated primary keys could also be omitted from this struct entirely.

This would integrate the Rust type checker with your insert query ensuring it will be valid at compile time.

Eg:
```rust
let new_cake = cake::InputModel {
  // id: does not exist on this struct because it is an auto-incremented column
  name: "Fruit cake".to_string(), // Required column because non-nullable and does not have a default value
  difficulty: None, // Not required because it has a default value of 0
};
Cake::insert(new_cake).exec(db).await?;
```

The main use-case for this is integrating with async-graphql.

**Comment by billy1624** [comment]
I think this is a good idea that utilizing Rust type checker to enforce nullability on compile time.

However, how can we derive the `InputModel` with a procedural macro?

Can we somehow read nullability of each column from the `ColumnDef` in procedural macro?

```rust
impl ColumnTrait for Column {
    type EntityName = Entity;

    fn def(&self) -> ColumnDef {
        match self {
            Self::Id => ColumnType::Integer.def(),
            Self::Name => ColumnType::String(None).def().null(),
        }
    }
}
```

**Reply by tqwewe** [reply]
That's also what I was investigating today.
I was hoping to do this from the ColumnTrait impl like you showed above, but I'm not sure how to make this work at compile time.

What I ended up doing for now is attributes and a `DeriveInputModel`:
```rust
#[DeriveModel, DeriveActiveModel, DeriveInputModel]
struct Model {
  #[sea(skip)]
  pub id: i64,
  pub name: String,
  #[sea(optional)]
  pub age: i64 // defaults to 10 in the database for example
}
```
This works at compile time.

Still, would be nice to be able to read it from the ColumnTrait def... but I'm not sure how possible it is.

**Reply by tqwewe** [reply]
@billy1624 another approach would be to have the cli generate a schema.rs file which contains the whole database schema as Rust types. This would allow it to be used for compile time macros.

**Comment by tyt2y3** [comment]
I think the central idea here is to use different models for different APIs.
`ActiveModel` to be used with  `save`, where the semantic is 'insert if primary key is unset, update otherwise'.
And to have specialized models to be used for `Insert` and `Update` respectively,
where their field definitions would help linting a correct query.

**Comment by shravanshetty1** [comment]


**Reply by tyt2y3** [reply]
What query are you doing and what errors did you get? Without more details we can't proceed.

**Reply by shravanshetty1** [reply]
https://github.com/SeaQL/sea-orm/discussions/98 - this describes what exactly happened

**Comment by 28Smiles** [comment]
While building my app, i encountered an error during insert, which could have been prevented by this at compile time. No I want to share my thoughts about this issue:
First off I am a big fan of pushing as much error as possible to the compile time. Since the cli already emits "non_null", another "default" annotation should not be an issue. While thinking about a possible InputModel, but it occurred to me, that just deriving another model would not fix this, since using the already established syntax of:
```
model::ActiveModel {
  value: 0,
  ..Default::default()
}
```
May possibly insert a default value for types which implement it, which may not be intentional. So i was thinking of the possibility of implementing the ActiveModelStruct as a builder using (const) generics. Using this one is able to track all fields set and unset. The derive macro can then implement an `IntoUpdateActiveModel` and `IntoInsertActiveModel` based on the state of the (const) generics. This is currently the only possibility I know of, but there may be issues with implementing the same trait twice, and therefore the macro may need to implement the traits for each set of generics. But, such an approach allows for typechecking for insert (all required fields/columns) and updates (primary key).

I hope this was helpful. I am very interested in your opinions and insights.
